<a href="https://colab.research.google.com/github/buildwithdemis/machinelearning/blob/main/ML_ch_09.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Exercise: Titanic Dataset - One-Hot Vectors**

**Categorical data aren't numerical**

Categorical data doesn't work with numbers the same way that other datatypes work with numbers. With ordinal or continuous (numerical) data, higher values imply an increase in amount. For example, a Titanic ticket price of £30 is more money than a ticket price of £12.

In contrast, categorical data has no logical order. We'll have problems if we try to encode, as numbers, categorical features that have more than two classes.

In this unit, we'll build a model to predict who survived the Titanic disaster. We'll practice transforming data between numerical and categorical types, including use of one-hot vectors.

**Data prepartion**

First, we'll open and quickly clean our dataset, like we did in the last unit:

In [9]:
import pandas
!wget https://raw.githubusercontent.com/MicrosoftDocs/mslearn-introduction-to-machine-learning/main/m0c_logistic_regression.py
!wget https://raw.githubusercontent.com/MicrosoftDocs/mslearn-introduction-to-machine-learning/main/Data/titanic.csv

# Open our dataset from file
dataset = pandas.read_csv("titanic.csv", index_col=False, sep=",", header=0)

# Fill missing cabin information with 'Unknown'
dataset["Cabin"].fillna("Unknown", inplace=True)

# Remove rows missing Age information
dataset.dropna(subset=["Age"], inplace=True)

# Remove the Name, PassengerId, and Ticket fields
# This is optional; it makes it easier to read our print-outs
dataset.drop(["PassengerId", "Name", "Ticket"], axis=1, inplace=True)

dataset.head()


--2025-06-02 23:12:31--  https://raw.githubusercontent.com/MicrosoftDocs/mslearn-introduction-to-machine-learning/main/m0c_logistic_regression.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 827 [text/plain]
Saving to: ‘m0c_logistic_regression.py.1’

m0c_logistic_regres 100%[===================>]     827  --.-KB/s    in 0s      

2025-06-02 23:12:32 (27.2 MB/s) - ‘m0c_logistic_regression.py.1’ saved [827/827]

--2025-06-02 23:12:32--  https://raw.githubusercontent.com/MicrosoftDocs/mslearn-introduction-to-machine-learning/main/Data/titanic.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... c

<ipython-input-9-4cf81e56779e>:9: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  dataset["Cabin"].fillna("Unknown", inplace=True)


Survived  Pclass     Sex   Age  SibSp  Parch     Fare    Cabin Embarked
0         0       3    male  22.0      1      0   7.2500  Unknown        S
1         1       1  female  38.0      1      0  71.2833      C85        C
2         1       3  female  26.0      0      0   7.9250  Unknown        S
3         1       1  female  35.0      1      0  53.1000     C123        S
4         0       3    male  35.0      0      0   8.0500  Unknown        S

**About Our Model**

We'll use a model training type known as Logistic Regression, which predicts who survived the Titanic disaster.

For this exercise, you don't need to understand logistic regression. We placed the implementation outside this notebook in a method called train_logistic_regression. If you're curious, you can read this method in our GitHub repository.

**The train_logistic_regression method:**

Accepts our data frame, and a list of features, to include in the model.
Trains the model.

Returns a number that states how well the model performs as it predicts passenger survival. Smaller numbers are better.

**Numerical Only**

Let's create a model that uses only the numerical features.

First, we'll use Pclass as an ordinal feature, rather than a one-hot encoded categorical feature.

In [10]:
from m0c_logistic_regression import train_logistic_regression

features = ["Age", "Pclass", "SibSp", "Parch", "Fare"]
loss_numerical_only = train_logistic_regression(dataset, features)

print(f"Numerical-Only, Log-Loss (cost): {loss_numerical_only}")

Numerical-Only, Log-Loss (cost): 0.6121682789483441


We have our starting point. Let's see if categorical features will improve the model.

**Binary Categorical Features**

Categorical features with only two potential values; for example, 0 or 1 - can be encoded in a single column.

We'll convert Sex values into IsFemale—a 0 for male and 1 for female—and include that in our model.

In [11]:
# Swap male / female with numerical values
# We can do this because there are only two categories
dataset["IsFemale"] = dataset.Sex.replace({'male':0, 'female':1})

# Print out the first few rows of the dataset
print(dataset.head())

# Run and test the model, also using IsFemale this time
features = ["Age", "Pclass", "SibSp", "Parch", "Fare", "IsFemale"]
loss_binary_categoricals = train_logistic_regression(dataset, features)

print(f"\nNumerical + Sex, Log-Loss (cost): {loss_binary_categoricals}")

   Survived  Pclass     Sex   Age  SibSp  Parch     Fare    Cabin Embarked  \
0         0       3    male  22.0      1      0   7.2500  Unknown        S   
1         1       1  female  38.0      1      0  71.2833      C85        C   
2         1       3  female  26.0      0      0   7.9250  Unknown        S   
3         1       1  female  35.0      1      0  53.1000     C123        S   
4         0       3    male  35.0      0      0   8.0500  Unknown        S   

   IsFemale  
0         0  
1         1  
2         1  
3         1  
4         0  

Numerical + Sex, Log-Loss (cost): 0.4707118595966243


<ipython-input-11-95afbae253e1>:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  dataset["IsFemale"] = dataset.Sex.replace({'male':0, 'female':1})


Our loss (error) decreased! This model performs better than the previous model.

**One-Hot Encoding**

Ticket class (Pclass) is an Ordinal feature. Its potential values (1, 2, and 3) have an order, and they have equal spacing. However, this even spacing might be incorrect. In stories about the Titanic, the third-class passengers were treated much worse than those in 1st and 2nd class.

Let's convert Pclass into a categorical feature with one-hot encoding:

In [12]:
# Get all possible categories for the "PClass" column
print(f"Possible values for PClass: {dataset['Pclass'].unique()}")

# Use Pandas to One-Hot encode the PClass category
dataset_with_one_hot = pandas.get_dummies(dataset, columns=["Pclass"], drop_first=False)

# Add back in the old Pclass column, for learning purposes
dataset_with_one_hot["Pclass"] = dataset.Pclass

# Print out the first few rows
dataset_with_one_hot.head()

Possible values for PClass: [3 1 2]


Survived     Sex   Age  SibSp  Parch     Fare    Cabin Embarked  IsFemale  \
0         0    male  22.0      1      0   7.2500  Unknown        S         0   
1         1  female  38.0      1      0  71.2833      C85        C         1   
2         1  female  26.0      0      0   7.9250  Unknown        S         1   
3         1  female  35.0      1      0  53.1000     C123        S         1   
4         0    male  35.0      0      0   8.0500  Unknown        S         0   

   Pclass_1  Pclass_2  Pclass_3  Pclass  
0     False     False      True       3  
1      True     False     False       1  
2     False     False      True       3  
3      True     False     False       1  
4     False     False      True       3

Note that the original Pclass data became three values: Pclass_1, Pclass_2 and Pclass_3.

Rows with Pclass values of 1 have a value in the Pclass_1 column. We see the same pattern for values of 2 and 3.

Now, re-run the model, and treat Pclass values as categorical values, instead of ordinal values.

In [13]:
# Run and test the model, also using Pclass as a categorical feature this time
features = ["Age", "SibSp", "Parch", "Fare", "IsFemale",
            "Pclass_1", "Pclass_2", "Pclass_3"]

loss_pclass_categorical = train_logistic_regression(dataset_with_one_hot, features)

print(f"\nNumerical, Sex, Categorical Pclass, Log-Loss (cost): {loss_pclass_categorical}")


Numerical, Sex, Categorical Pclass, Log-Loss (cost): 0.4717302923736227


This seems to have made things slightly worse!

Let's move on.

**Including Cabin**

Recall that many passengers had Cabin information. Cabin is a categorical feature and should be a good predictor of survival, because people in lower cabins probably had little time to escape during the sinking.

Let's encode cabin using one-hot vectors and include it in a model. This time, there are so many cabins that we won't print them all out. To practice printing them out, feel free to edit the code.

In [14]:
# Use Pandas to One-Hot encode the Cabin and Pclass categories
dataset_with_one_hot = pandas.get_dummies(dataset, columns=["Pclass", "Cabin"], drop_first=False)

# Find cabin column names
cabin_column_names = list(c for c in dataset_with_one_hot.columns if c.startswith("Cabin_"))

# Print out how many cabins there were
print(len(cabin_column_names), "cabins found")

# Make a list of features
features = ["Age", "SibSp", "Parch", "Fare", "IsFemale",
            "Pclass_1", "Pclass_2", "Pclass_3"] + \
            cabin_column_names

# Run the model and print the result
loss_cabin_categorical = train_logistic_regression(dataset_with_one_hot, features)

print(f"\nNumerical, Sex, Categorical Pclass, Cabin, Log-Loss (cost): {loss_cabin_categorical}")

135 cabins found

Numerical, Sex, Categorical Pclass, Cabin, Log-Loss (cost): 0.4600048688202762


**Improving Power**

Including very large numbers of categorical classes (for example, 135 cabins) is often not the best way to train a model. This is because the model has only a few examples of each category class to learn from.

Sometimes, we can improve models if we simplify features. Cabin probably helped because it indicated the Titanic deck where people were probably located. Those in lower decks would have had their quarters flooded first.

It might be simpler to use deck information instead of categorizing people into Cabins.

Let's simplify what we have run, replacing the 135 Cabin categories with a simpler Deck category that has only 9 values: A through G, T, and U (Unknown):

In [15]:
# We have cabin names, like A31, G45. The letter refers to the deck that
# the cabin was on. Extract just the deck and save it to a column.
dataset["Deck"] = [c[0] for c in dataset.Cabin]

print("Decks: ", sorted(dataset.Deck.unique()))

# Create one-hot vectors for:
# Pclass - the class of ticket. (This could be treated as ordinal or categorical)
# Deck - the deck that the cabin was on
dataset_with_one_hot = pandas.get_dummies(dataset, columns=["Pclass", "Deck"], drop_first=False)

# Find the deck names
deck_of_cabin_column_names = list(c for c in dataset_with_one_hot.columns if c.startswith("Deck_"))

features = ["Age", "IsFemale", "SibSp", "Parch", "Fare",
            "Pclass_1", "Pclass_2", "Pclass_3",
            "Deck_A", "Deck_B", "Deck_C", "Deck_D",
            "Deck_E", "Deck_F", "Deck_G", "Deck_U", "Deck_T"]

loss_deck = train_logistic_regression(dataset_with_one_hot, features)

print(f"\nSimplifying Cabin Into Deck, Log-Loss (cost): {loss_deck}")

Decks:  ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'T', 'U']

Simplifying Cabin Into Deck, Log-Loss (cost): 0.4588927791855666


**Comparing Models**

Let's compare the loss for these models:

In [16]:
# Use a dataframe to create a comparison table of metrics
# Copy metrics from previous Unit

l =[["Numeric Features Only", loss_numerical_only],
    ["Adding Sex as Binary", loss_binary_categoricals],
    ["Treating Pclass as Categorical", loss_pclass_categorical],
    ["Using Cabin as Categorical", loss_cabin_categorical],
    ["Using Deck rather than Cabin", loss_deck]]

pandas.DataFrame(l, columns=["Dataset", "Log-Loss (Lower is better)"])

Dataset  Log-Loss (Lower is better)
0           Numeric Features Only                    0.612168
1            Adding Sex as Binary                    0.470712
2  Treating Pclass as Categorical                    0.471730
3      Using Cabin as Categorical                    0.460005
4    Using Deck rather than Cabin                    0.458893

If we include categorical features, we can both improve and harm how well a model works. Often, experimentation is the best way to find the best model.

**Summary**

In this unit you learned how to use one-hot encoding to address categorical data.

We also explored how critical thinking about the problem at hand can sometimes improve a solution more than simply including all possible features in a model.

**Brainstorming **


*   Why do we clean our data before training?
*   What kind of data are best encoded with one-hot vectors?
      Categorical data with three or more values
*   What is a data sample? What is a population?
*   You have a model that doesn't perform well. Which of these options definitely do not help improve its performance?



